File: hydro/src/main/java/com/ultreon/hydro/graphics/shapes/Polygon.java
package com.ultreon.hydro.graphics.shapes;

import com.ultreon.commons.util.Constants;
import com.ultreon.hydro.util.IntersectionUtils;

import java.util.ArrayList;

public class Polygon implements Shape {
    private Point[] points;

    public Polygon(Point[] points) {
        this.points = points;
    }

    public Polygon(java.awt.Polygon polygon) {
        int[] xps = polygon.xpoints;
        int[] yps = polygon.ypoints;
        int np = polygon.npoints;

        ArrayList<Point> points1 = new ArrayList<>();
        Point[] points2 = new Point[]{};

        for (int i = 0; i < np; i++) {
            int xp = xps[i];
            int yp = yps[i];
            points1.add(new Point(xp, yp));
        }

        points1.toArray(points2);
        this.points = points2;
    }

    /**
     * @return the points
     */
    public Point[] getPoints() {
        return points;
    }

    /**
     * @param points the points to set
     */
    public void setPoints(Point[] points) {
        this.points = points;
    }

    @Override
    public boolean doIntersect(Shape shape) {
        if (shape instanceof Circle)
            return IntersectionUtils.doIntersect((Circle) shape, this);
        else if (shape instanceof Line)
            return IntersectionUtils.doIntersect(this, (Line) shape);
        else
            throw new UnsupportedOperationException(Constants.UNSUPPORTED_SHAPE);
    }
}

File: src/main/kotlin/qtech/bubbles/common/graphics/shapes/Polygon.kt
package qtech.bubbles.common.graphics.shapes

import qtech.bubbles.core.utils.categories.Constants
import qtech.bubbles.core.utils.categories.IntersectionUtils
import java.awt.Polygon

class Polygon : Shape {
    var points: Array<Point>

    constructor(points: Array<Point>) {
        this.points = points
    }

    constructor(polygon: Polygon) {
        val xps = polygon.xpoints
        val yps = polygon.ypoints
        val np = polygon.npoints
        val points1 = ArrayList<Point>()
        val points2 = arrayOf<Point>()
        for (i in 0 until np) {
            val xp = xps[i]
            val yp = yps[i]
            points1.add(Point(xp.toDouble(), yp.toDouble()))
        }
        points1.toArray(points2)
        points = points2
    }

    override fun doIntersect(shape: Shape): Boolean {
        return when (shape) {
            is Circle -> IntersectionUtils.doIntersect(
                shape,
                this
            )
            is Line -> IntersectionUtils.doIntersect(
                this,
                shape
            )
            else -> throw UnsupportedOperationException(Constants.UNSUPPORTED_SHAPE)
        }
    }
}